### ISSUE TITLE
Why is Sequence[T] invariant? (difference with mypy)

### ISSUE THREAD

**Issue by gvanrossum-ms** [microsoft/pyright#10048]
I feel embarrassed that my brain has atrophied to the point where I have to file this issue. Please explain what I am missing.

I have a very simple example.
```py
class IMessage:
    pass

class IMessageList:
    messages: tuple[IMessage]

class MyMessage(IMessage):
    pass

class MyMessageList(IMessageList):
    messages: tuple[MyMessage]
```
While mypy 1.15.0 approves of this, pyright 1.1.396 complains on the very last line of the example that messages is invariant and tuple[MyMessage] is not compatible with tuple[IMessage].
```
/home/gvanrossum/x.py
  /home/gvanrossum/x.py:11:5 - error: "messages" overrides symbol of same name in class "IMessageList"
    Variable is mutable so its type is invariant
      Override type "tuple[MyMessage]" is not the same as base type "tuple[IMessage]" (reportIncompatibleVariableOverride)
1 error, 0 warnings, 0 informations 
```

Note, I tried the same with Sequence instead of tuple, and it gives the same error. Also with tuple[T, ...], which I meant to use.

**Comment by erictraut** [comment]
The variable's type is invariant because the variable is mutable. That means anyone acting upon the base class (including any method within the base class) is free to assign a new value to the variable that does not honor the type of the subclass.

```python
class IMessage:
    pass

class IMessageList:
    messages: tuple[IMessage]

    def reset(self):
        self.messages = (IMessage(),)

class MyMessage(IMessage):
    x: int

class MyMessageList(IMessageList):
    messages: tuple[MyMessage]

    def print_x(self):
        print(self.messages[0].x)

m = MyMessageList()
m.messages = (MyMessage(),)

m.print_x()  # OK
m.reset()
m.print_x()  # Crash
```

There is a draft [PEP 767](https://peps.python.org/pep-0767/) under development that would allow an instance variable to be marked "read only". A read-only attribute can be safely overridden in a covariant manner.

Mypy recently added an option to enforce invariance for mutable attributes, but it's not (currently) enabled by default because the maintainers were concerned about the compatibility impact. I think they're considering making this the default behavior for the next major release of mypy.

**Comment by gvanrossum-ms** [comment]
Sadly, I will probably end up adding a `# type: ignore`, since in my environment I have a use for this particular inheritance pattern and I am not in the least worried about such a reset() method. Sorry. (And hi!)

**Comment by erictraut** [comment]
One way to make this type safe is to use generics.

```python
class IMessage:
    pass

class IMessageList[T: IMessage]:
    messages: tuple[T]

class MyMessage(IMessage):
    pass

class MyMessageList(IMessageList[MyMessage]):
    pass
```

**Comment by jorenham** [comment]
Another option is to use properties:

```py
class IMessage:
    pass

class IMessageList:
    @property
    def messages(self) -> tuple[IMessage]: ...

class MyMessage(IMessage):
    pass

class MyMessageList(IMessageList):
    @property
    def messages(self) -> tuple[MyMessage]: ...
```

**Comment by gvanrossum-ms** [comment]
> One way to make this type safe is to use generics.

That's **exactly** the solution I came up with!
